[#1089] Way of working quality (#1075)
kiesraad/e-ks: docs/werkwijze-kwaliteit-waarborgen.md

# Werkwijze e-KS team: kwaliteit waarborgen

Dit document beschrijft de werkwijze van het e-KS team en hoe er kwaliteit wordt gewaarborgd. Kwaliteit is hierbij geen aparte stap aan het einde van het ontwikkelproces, maar een geïntegreerd onderdeel van de werkwijze. 

> Er is niet alleen aandacht voor de technische kwaliteit, maar ook of de gebouwde functionaliteit voldoet aan wetgeving en de behoeftes ondersteunt. 

Reviewer: ```suggestion
Er is aandacht voor de technische kwaliteit en functionele kwaliteit. Functionele kwaliteit wordt gewaarborgd door de applicatie te toetsen aan wetgeving en gebruikersbehoeftes.
```

## Ontwikkeling

### Algemene werkwijze
- Er wordt gezorgd voor continue kwaliteit door **iteratief te ontwikkelen**:

>   - Het e-KS team werkt in twee wekelijkse sprints, inclusief alle sprintevents.  

Reviewer: ```suggestion
  - Het e-KS werkt [agile](https://agilemanifesto.org/) door middel van [Scrum](https://www.scrum.org/)
  - Het e-KS team werkt in twee wekelijkse sprints, inclusief alle Scrum ceremonies.
```

>   - Refinement wordt gedaan vlak voordat er wordt begonnen met de ontwikkeling van nieuwe functionaliteit

Reviewer: ```suggestion
  - Refinement wordt deels gedaan tijdens de sprint planning en deels ad hoc voordat er wordt begonnen met de ontwikkeling van nieuwe functionaliteit
```

  - De Product Owner stemt zowel voor als na de ontwikkeling de functionaliteit af met interne stakeholders

>   - Playwright tests worden gelijktijdig met de ontwikkeling geschreven

Reviewer: ```suggestion
  - Systeem testen worden gelijktijdig met de ontwikkeling geïmplementeerd met [Playwright](https://playwright.dev/)
```

>   - Na elke pull request worden de CI/CD pipeline, met quality gates, afgetrapt

Reviewer: ```suggestion
  - Nieuwe features worden ingediend door middel van Pull Requests
  - Pull Requests worden pas geaccepteerd in de main branch zodra:
     - De CI/CD pipeline, met quality gates, slaagt
     - De wijziging door minstens één andere ontwikkelaar is gereviewd en goedgekeurd (vierogenprincipe)
```

- Er wordt gezorgd voor een gezamenlijk en objectief kwaliteitsniveau door de **Definition of Ready (DoR) en Definition of Done (DoD)**:
  - DoR: issue mag alleen worden opgepakt als er is voldaan aan de DoR [link]

>   - DoD: issue is pas klaar als er is voldaan aan de DoD [link]

Reviewer: ```suggestion
  - DoR: issue mag alleen worden opgepakt als er is voldaan aan de [DoR](docs/definition-of-ready.md)
  - DoD: issue is pas klaar als er is voldaan aan de [DoD](docs/definition-of-done.md)
```

- Kwaliteit wordt verhoogd door het **vierogenprincipe**:

>   - Elke pull request wordt gereviewed door een andere ontwikkelaar voordat code gemerged mag worden

Reviewer: ```suggestion
```

 
### Continuous Integration / Continuous Delivery (CI/CD)

CI/CD  helpt om risico’s  te mitigeren. Door codewijzigingen regelmatig (bij elke PR) te integreren, testen en opleveren, worden fouten eerder ontdekt en wordt de kans op problemen bij het mergen en opleveren van nieuwe code verkleind. Door een CI/CD-pipeline in te richten kunnen deze stappen geautomatiseerd worden gedaan.

De CI/CD pipeline binnen e-KS bevat de volgende quality gates:
- Testing
  - Alle playwright tests moeten slagen
  - Alle unit testen moeten slagen
  - Er moet een unit test coverage van >80% zijn op nieuwe code
- Code style
  - Rust format geeft geen warning
  - Clippy geeft geen warnings
  - Geen Biome linting errors op CSS en TypeScript code
  - Geen djlint format warnings op HTML templates
  - De CI/CD pipeline configuratie mag geen zizmor linting errors bevatten

>   - Elke bestand in het archief moet eindigen met een lijntje met een een newline ('\n') karakter

Reviewer: ```suggestion
  - Elke bestand in het archief moet eindigen met een witregel
```
